### ISSUE TITLE
Implement Agent Interface and DQN Agent Template

### ISSUE THREAD

**Issue by seekersoftec** [fortesenselabs/trade_flow#5]
**Task:**

1. **Create an Agent Interface:** Define a generic interface that outlines the common methods and properties required by different types of reinforcement learning agents.
2. **Create a DQN Agent Template:** Provide a template for implementing a Deep Q-Network (DQN) agent, following the standard DQN architecture and algorithm, using PyTorch.

**Agent Interface (Python):**

```python
class Agent:
    def act(self, state):
        pass

    def learn(self, state, action, reward, next_state, done):
        pass

    def save(self, filename):
        pass

    def load(self, filename):
        pass

    # Other common agent methods
```

**DQN Agent Template (PyTorch):**

```python
import torch
import torch.nn as nn
import torch.optim as optim

class DQNAgent(Agent):
    def __init__(self, state_dim, action_dim, learning_rate=0.01, gamma=0.99, epsilon=1.0, epsilon_min=0.01, epsilon_decay=0.995):
        # Initialize DQN network
        self.model = self.build_model(state_dim, action_dim)
        self.target_model = self.build_model(state_dim, action_dim)
        self.target_model.load_state_dict(self.model.state_dict())
        self.target_model.eval()

        # Set hyperparameters
        self.learning_rate = learning_rate
        self.gamma = gamma
        self.epsilon = epsilon
        self.epsilon_min = epsilon_min
        self.epsilon_decay = epsilon_decay

        self.optimizer = optim.Adam(self.model.parameters(), lr=learning_rate)

    def build_model(self, state_dim, action_dim):
        # Build the DQN neural network architecture
        model = nn.Sequential(
            nn.Linear(state_dim, 256),
            nn.ReLU(),
            nn.Linear(256, 128),
            nn.ReLU(),
            nn.Linear(128, action_dim)
        )
        return model

    def act(self, state):
        # Choose action based on epsilon-greedy policy
        if np.random.rand() <= self.epsilon:
            return np.random.randint(action_dim)
        else:
            state = torch.tensor(state, dtype=torch.float32)
            with torch.no_grad():
                q_values = self.model(state)
            return q_values.max(1)[1].item()

    def learn(self, state, action, reward, next_state, done):
        # Update DQN network using experience replay
        # ...

    # Implement other methods (e.g., save, load, update target model)
```

**Additional Considerations:**

* **Experience Replay:** Implement a replay buffer to store past experiences and sample them randomly for training.
* **Target Network:** Use a target network to stabilize training by providing a fixed target Q-value during updates.
* **Exploration-Exploitation:** Balance exploration (trying new actions) and exploitation (choosing the best known action) using techniques like epsilon-greedy.
* **Neural Network Architecture:** Experiment with different neural network architectures (e.g., convolutional neural networks, recurrent neural networks) based on the problem domain.
* **Hyperparameter Tuning:** Tune hyperparameters like learning rate, discount factor, and exploration rate to optimize agent performance.

By providing an agent interface and a DQN agent template using PyTorch, we can establish a framework for creating and experimenting with different reinforcement learning agents, making it easier to develop and test trading strategies.
